src/server/db/actions/GameAction.ts

```diff
@@ -72,70 +72,124 @@ export async function editGame(allData: nextEditGame) {
-      new: false,
+  //Find the correspoding tag ids for the accessibility TAGS.
+  let accessibility: ObjectId[] = [];
+  if (query.accessibility) {
+    const accessibilityTags = await TagModel.find({
+      name: { $in: query.accessibility },
+      type: "accessibility",
     });
-    if (!result) {
-      throw new GenericUserErrorException("Game with given ID does not exist.");
+    //Ensure that we found all tags and no extras
+    if (query.accessibility.length !== accessibilityTags.length) {
+      throw new GenericGameErrorException(
+        "One or more of the accessibility tags do not exist.",
+      );
     }
-    return result;
-  } catch (e) {
-    throw e;
+    accessibility = accessibilityTags.map((tag) => tag.id);
   }
-}
+  let custom: ObjectId[] = [];
+  if (query.tags) {
+    const normalTags = await TagModel.find({
+      name: { $in: query.tags },
+      type: "custom",
+    });
+    if (query.tags.length !== normalTags.length) {
+      throw new GenericGameErrorException(
+        "One or more of the custom tags do not exist.",
+      );
+    }
+    custom = normalTags.map((tag) => tag.id);
+  }
+  const totalTags = [...custom, ...accessibility];
+  if (totalTags.length !== 0) {
+    //Add to filters
+    filters.tags = { $all: totalTags };
+  }
+  //Currently ignoring filtering by build, as that hasn't yet been implemented.
 
-export async function getAllGames() {
-  await connectMongoDB();
-  try {
-    const games = await GameModel.find();
-    if (games == null) {
-      return [];
+  //Match by name with regex
+  if (query.name && query.name !== "") {
+    const reg_string = new RegExp(query.name, "i");
+    filters.name = { $regex: reg_string };
+  }
+  //Match by gameContent, by ensuring they all exist.
+  if (query.gameContent) {
+    filters = query.gameContent.reduce((acc, curr) => {
+      acc[curr] = { $exists: true };
+      return acc;
+    }, filters);
+  }
+
+  //Filtering based on game build (filtering should be add)
+  if (query.gameBuilds) {
+    if (query.gameBuilds.includes("webgl")) {
+      //WebGl requires a different type of filter, as it is seperate from the other builds
+      filters.webGLBuild = true;
+      //Remove webGL for further filtering based on other normal games.
+      query.gameBuilds = query.gameBuilds.filter((item) => item !== "webgl");
     }
-    return games;
-  } catch (e) {
-    throw e;
+    filters.builds = { $all: query.gameBuilds };
   }
+
+  let games = await GameModel.find(filters)
+    .skip((query.page - 1) * RESULTS_PER_PAGE)
+    .limit(RESULTS_PER_PAGE);
+
+  return games;
 }
 
 export async function getGameById(id: string) {
   await connectMongoDB();
-  try {
-    const game = await GameModel.findById(id)
-      .populate("themes")
-      .populate("tags");
-    return game;
-  } catch (e) {
-    throw e;
-  }
+  const game = await GameModel.findById(id).populate<IGame>("themes tags");
```

Comment:
The type passed via `.populate()` should represent the types of the objects that are populated:

```js
.populate<ITheme>("themes").populate<ITag>("tags")
```